tfhe/src/integer/ciphertext/compact_list.rs

```diff
@@ -1,42 +1,380 @@
+}
+
+impl CompactCiphertextListExpander {
+    fn new(expanded_blocks: Vec<Ciphertext>, info: Vec<DataKind>) -> Self {
+        Self {
+            expanded_blocks,
+            info,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.info.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn blocks_of(&self, index: usize) -> Option<(&[Ciphertext], DataKind)> {
+        let preceding_infos = self.info.get(..index)?;
+        let current_info = self.info.get(index).copied()?;
+
+        let start_block_index = preceding_infos
+            .iter()
+            .copied()
+            .map(DataKind::num_blocks)
+            .sum();
+        let end_block_index = start_block_index + current_info.num_blocks();
+
+        self.expanded_blocks
+            .get(start_block_index..end_block_index)
+            .map(|block| (block, current_info))
+    }
+
+    pub fn get_kind_of(&self, index: usize) -> Option<DataKind> {
+        self.info.get(index).copied()
+    }
+
+    pub fn get<T>(&self, index: usize) -> Option<crate::Result<T>>
+    where
+        T: Expandable,
+    {
+        self.blocks_of(index)
+            .map(|(blocks, kind)| T::from_expanded_blocks(blocks, kind))
+    }
+
+    pub(crate) fn message_modulus(&self) -> MessageModulus {
+        self.expanded_blocks[0].message_modulus
+    }
+}
+
 #[derive(Clone, Serialize, Deserialize)]
 pub struct CompactCiphertextList {
     pub(crate) ct_list: crate::shortint::ciphertext::CompactCiphertextList,
-    // Keep track of the num_blocks, as we allow
-    // storing many integer that have the same num_blocks
-    // into ct_list
-    pub(crate) num_blocks_per_integer: usize,
+    // Integers stored can have a heterogeneous number of blocks and signedness
+    // We store this info to safeguard the expansion
+    pub(crate) info: Vec<DataKind>,
 }
 
 impl ParameterSetConformant for CompactCiphertextList {
-    type ParameterSet = RadixCompactCiphertextListConformanceParams;
+    type ParameterSet = CompactCiphertextListConformanceParams;
+
+    fn is_conformant(&self, params: &CompactCiphertextListConformanceParams) -> bool {
+        if !params.num_elements_constraint.is_valid(self.info.len()) {
+            return false;
+        }
 
-    fn is_conformant(&self, params: &RadixCompactCiphertextListConformanceParams) -> bool {
-        self.num_blocks_per_integer == params.num_blocks_per_integer
-            && self
-                .ct_list
-                .is_conformant(&params.to_shortint_ct_list_conformance_parameters())
+        self.is_conformant_with_shortint_params(params.shortint_params)
     }
 }
 
 impl CompactCiphertextList {
-    pub fn expand_one<T: IntegerRadixCiphertext>(&self) -> T {
-        let mut blocks = self.ct_list.expand();
-        blocks.truncate(self.num_blocks_per_integer);
-        T::from(blocks)
+    pub fn is_packed(&self) -> bool {
+        self.ct_list.degree.get() > self.ct_list.message_modulus.0
```

Comment:
properly set at encryption ? I believe some encryption apply the message modulus on the inputs of blocks

Reply:
Yes it is, In the test in integer::ciphertext::public_key::test we test both the case of building the list packed and unpacked

I will add an explicit call to this function in the test tho